### ISSUE TITLE
Support for Custom domains for Web App

### ISSUE THREAD

**Issue by jostein-skaar** [CompositionalIT/farmer#213]
It would be nice to be able to add mydomain.com, and even better the free ssl certificate part as well. (Luckily it doesn't get removed if one redeploy.)

Be aware: Custom domains are not available for the free tier.

I guess this is some documentation on the matter: 
https://github.com/Azure/azure-quickstart-templates/blob/master/201-web-app-custom-domain-and-ssl/azuredeploy.json

**Comment by michaelkc** [comment]
AFAIK the (App Service Managed) certificate part currently suffers from a chicken-and-the-egg problem: The certificate generation cannot be specified in Arm, and cannot be done without a running App Service with a custom domain configured for the certificate subject. So you basically have to define the App Service with custom domain in Arm:
```json 
{
      "type": "Microsoft.Web/sites/hostNameBindings",
      "name": "[concat(parameters('appServiceName'),'/',parameters('cname'))]",
      "apiVersion": "2016-08-01",
      "location": "[parameters('location')]",
      "scale": null,
      "properties": {
        "siteName": "[parameters('cname')]",
        "domainId": null,
        "hostNameType": "Verified"
      },
      "dependsOn": [
        "[resourceId('Microsoft.Web/sites', parameters('appServiceName'))]"
      ]
}
```
Deploy it, then generate the certificate
```bash
az webapp config ssl create --resource-group "someresourcegroup" --name "someappservice" --hostname "somecname.example.com"
```
extract the thumbprint and finally bind the certificate to the app service, something like
```json 
{
  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {},
  "variables": {},
  "resources": [
    {
      "type": "Microsoft.Web/sites",
      "kind": "app",
      "name": "someappservice",
      "apiVersion": "2016-08-01",
      "location": "westeurpe",
      "properties": {
        "hostNameSslStates": [
          {
            "name":"somecname.example.com",
            "sslState":"SniEnabled",
            "ipBasedSslResult":null,
            "virtualIP":null,
            "thumbprint":"<thumbprint>",
            "toUpdate":true,
            "toUpdateIpBasedSsl":null,
            "iPBasedSslState":"NotConfigured",
            "hostType":"Standard"
          }
        ],
      "httpsOnly": true
      },
      "dependsOn": []
    }
  ]
}  
```
(the last step can also be done via `az`, but it has known bugs which manifest when the App Service Plan and App Service are in separate resource groups - the Arm binding above works around that)
To summarize, I do not think a pure Arm solution is currently possible for App Service Managed Certificates. 

They are still in preview though, so that might change eventually.

**Comment by isaacabraham** [comment]
We could look at doing this with a post-deploy task - Farmer already supports this for things like uploading code into a webapp. However, it sounds like this one would be even more involved because it would need ARM deploy, Az call and then *another* ARM deploy!

**Comment by BrianVallelunga** [comment]
I think this can be implemented as of today, but it may need multiple ARM deployments. The first thing that would be needed is to support `hostnameBindings` creation in the builder for web apps and functions apps. I didn't see this in the API, but maybe I missed it.

Then, in a nested template, the managed certificate would need to be created. I'm not exactly sure how nested templates work, but here's an example of an ARM template that supposedly creates the certificate.

    "resources": [
        {
            "apiVersion": "2019-08-01",
            "location": "[parameters('Location')]",
            "name": "[concat(parameters('SubjectName'), '-', parameters('AppServicePlanName'), '-', parameters('Location'))]",
            "type": "Microsoft.Web/certificates", 
            "properties": {
                "serverFarmId": "[resourceId('Microsoft.Web/serverfarms/', parameters('AppServicePlanName'))]",
                "canonicalName": "[parameters('SubjectName')]",
                "domainValidationMethod":"http-token"
            }
        }
      ]

More references are here:
* https://gist.github.com/NickDarvey/814f0f34c03dff9f8fbe56310f0bef28
* https://dotnetdevlife.wordpress.com/2019/11/12/arm-app-service-managed-certificate/
* https://azure.github.io/AppService/2021/03/02/asmc-apex-domain.html
* https://github.com/MicrosoftDocs/azure-docs/issues/42539

**Comment by TheRSP** [comment]
I believe this should now be possible with a nested deployment. In the outer deployment, you create the app, hostname binding and certificate and then, in a nested deployment, you update the hostname binding to reference the certificate.

 I believe @Iamabdul is very close to having a prototype implementation of this.
